foolbox/batch_attacks/virtual_adversarial.py

```diff
@@ -0,0 +1,122 @@
+        iterations : int
+            Number of iterations to perform power method to search for second
+            order perturbation of KL divergence.
+        epsilons : int or Iterable[float]
+            Either Iterable of step sizes in the direction of the sign of
+            the gradient or number of step sizes between 0 and max_epsilon
+            that should be tried.
+        max_epsilon : float
+            Largest step size if epsilons is not an iterable.
+        """
+
+        assert a.target_class() is None, 'Virtual Adversarial is an ' \
+                                         'untargeted adversarial attack.'
+
+        yield from self._run(a, xi=xi, iterations=iterations,
+                             epsilons=epsilons, max_epsilon=max_epsilon)
+
+    def _run(self, a, xi, iterations, epsilons, max_epsilon):
+        if not a.has_gradient():
+            return
+
+        x = a.unperturbed
+        min_, max_ = a.bounds()
+
+        logits, _ = yield from a.forward_one(x)
+
+        if not isinstance(epsilons, Iterable):
+            epsilons = np.linspace(0, max_epsilon, num=epsilons + 1)[1:]
+            decrease_if_first = True
+        else:
+            decrease_if_first = False
+
+        for _ in range(2):  # to repeat with decreased epsilons if necessary
+            for i, epsilon in enumerate(epsilons):
+                # start with random vector as search vector
+                d = np.random.normal(0.0, 1.0, size=x.shape)
+                for it in range(iterations):
+                    # normalize proposal to be unit vector
+                    d = xi * d / np.sqrt((d**2).sum())
+
+                    logits_d, _ = yield from a.forward_one(x + d)
+
+                    dl_dp = softmax(logits) - softmax(logits_d)
+
+                    # d = dl_dd = dl_dp * dp_dd
+                    # use gradient of KL divergence as new search vector
+                    d = yield from a.backward_one(
+                        gradient=dl_dp,
+                        x=x + d, strict=False
+                    )
+                    d = (max_ - min_) * d
+
+                    if np.allclose(np.sqrt((d**2).sum()), 0, atol=1e-16):
+                        raise RuntimeError('Gradient vanished; this can happen '
+                                           'if xi is too small.')
+
+                delta = d / np.sqrt((d**2).mean())
+
+                perturbed = x + self._clip_perturbation(a, delta, epsilon)
+                perturbed = np.clip(perturbed, 0, 1)
+
+                _, is_adversarial = yield from a.forward_one(perturbed)
+                if is_adversarial:
+                    if decrease_if_first and i < 20:
+                        logging.info('repeating attack with smaller epsilons')
+                        break
+                    return
+
+            max_epsilon = epsilons[i]
+            epsilons = np.linspace(0, max_epsilon, num=20 + 1)[1:]
```

Comment:
what are these two lines for?

Reply:
The outer loop is not mentioned in the reference implementation or the paper but I added it nevertheless because the SingleStepGradientBaseAttack also does this tuning of epsilon. I am not sure if we should keep it or remove it so that the attack is closer to the original version. What do you think about it?